plonky3/src/circuit_builder.rs

```diff
@@ -28,17 +28,24 @@ impl<'a, T: FieldElement> PowdrCircuit<'a, T> {
     pub fn generate_trace_rows(&self) -> RowMajorMatrix<Goldilocks> {
         // an iterator over all columns, committed then fixed
         let witness = self.witness().iter();
+        let publics = self.get_publics().into_iter();
         let len = self.analyzed.degree.unwrap();
 
         // for each row, get the value of each column
         let values = (0..len)
             .flat_map(move |i| {
+                // witness values
                 witness
                     .clone()
                     .map(move |(_, v)| cast_to_goldilocks(v[i as usize]))
+                    .chain(
+                        publics
+                            .clone()
+                            .map(move |(_, _, idx)| T::from(i as usize == idx))
+                            .map(cast_to_goldilocks),
```

Comment:
Actually the cast to goldilocks can be done a single time at the end of the chain rather than twice now